### PR TITLE
First Fully Funtional Pendulum Example

### LINKED ISSUE #25: Milestone: Train a simple task from human feedback
# Our Goal

Our current focus should be to train some behavior end-to-end using our own feedback. The task can be very simple (Pendulum) and we can use imitation for the training.

The "user story" is the following:

- Start server
- Start imitation_example_client.py 
- The imitation example client starts training the agent, regularly sends queries to the server.
- The labeler responds to queries in their web browser (the query client).
- The imitation example client regularly pulls the results and retrains the reward model (imitation takes care of this if the Querent and the Gatherer are implemented correctly).

We can use this for a simple demo with the Pendulum environment. For this, we might need to adjust the hyper-parameters in our example (compare imitation#771). Training the agent should require no more than 200 queries, so (assuming 2s per feedback) < 10 minutes.

# Related issues and PRs

- [x] The RAM issue is a blocker #22.
- [ ] We should update the hyperparameters in the example #28.

### LINKED ISSUE #32: System Crash due to Cache Issues
The current state of work in https://github.com/timokau/prefq/pull/27 led to overflowing disk (cached data was written to disk) on my WSL System, which ended up blocking disk space permanently and could not be easily freed again. The issue lies in the `mkdtemp()` function (used in [updated_imitation_modules.py](https://github.com/timokau/prefq/blob/main/prefq/examples/updated_imitation_modules.py)), which is a C based library function, that does not offer any built-in mechanism to limit disk space usage and therefore requires the user to ensure in a bulletproof manner, that no memory leaks occur. This is a major disadvantage in terms of maintainability & usability. 

Under all tests using gymnasium 0.28.0 no errors occurred, it has just been now, after updating to gymnasium 0.29.0, that the execution of the [imitation example](https://github.com/timokau/prefq/blob/main/prefq/examples/imitation_preference_comparisons.py) led to a huge memory leak of roughly 70GB. I cannot fully exclude the possibility, that this has been a cumulative memory leak stacking up over several days of testing, however, to me it seems very likely that this is not the case, considering that a minimum of 500 video pairs has been evaluated on my machine, while a batch of 20 video pairs & subsequent training can already consume roughly 6GB of storage.

Also, it is important to mention, that the blocked disk space seemed (this statement is not fully verified, as 5GB of cached data, unrelated to PrefQ) to be available within WSL, and only appeared blocked within the Windows Filesystem. Either way, some type of security barrier must urgently be implemented.

In this light, considerations regarding a more manageable and risk-free solution, seem necessary. The optimal solution would be a cache object, that automatically clears all allocated memory, after either successful or forceful termination (SIGKILL w/ ctrl+c). The Cache Memory used in [updated_imitation_modules.py](https://github.com/timokau/prefq/blob/main/prefq/examples/updated_imitation_modules.py) appears to only be closed, when the `RenderImageInfoWrapper.close()` function is called, which triggers `shutil.rmtree(self.file_cache)`. This function might not be called anymore under the updated Gymnasium/Imitation API, possibly leading to the issue, that we are observing here. 

During a short search, i came across these two promising libraries:

https://pypi.org/project/diskcache/
https://pypi.org/project/cache-to-disk/

### PR DESCRIPTION
**Pull request by patrickab**
This Pull Request provides the first fully functional example, capable of learning the Pendulum Task without causing memory leaks.

- by integrating a Context Manager the issue described in https://github.com/timokau/prefq/issues/32 could be resolved. Fixes #32.
- by sorting the feedback received from the server, we ensure, that proper behavior is returned from the PrefqGatherer
- by using the hyperparameters proposed in https://github.com/HumanCompatibleAI/imitation/pull/771, the example is capable of learning the desired behavior much faster.

Fixes #25!

### CODE REVIEW

File: prefq/examples/imitation_preference_comparisons.py [hunk 1]
```diff
@@ -221,14 +221,20 @@ def _wait_for_feedback_request(url):
     ),
     env=venv,
     n_steps=2048 // venv.num_envs,
+    clip_range=0.1,
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:224]: Please add a comment to the top of the file that indicates why we chose these hyperparameters (a link to the PR in which I propose them should be fine).

File: prefq/examples/imitation_preference_comparisons.py [hunk 2]
```diff
@@ -145,11 +145,12 @@ def _send_videos_to_server(self, query_id):
...
-        if response.status_code >= 200 % response.status_code < 400:
+        try:
+            response.raise_for_status()
             print("PrefqGatherer: Payload transferred")
-        else:
-            print("PrefqGatherer: Error while sending videos")
-            print(response.status_code)
+        except requests.exceptions.HTTPError as err:
+            print(f"PrefqGatherer: Error while sending videos: {err}")
+            raise requests.exceptions.HTTPError
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:153]: You can just `raise err` to avoid losing information by re-creating the exception.

File: prefq/examples/imitation_preference_comparisons.py [hunk 3]
```diff
@@ -167,16 +168,23 @@ def _get_feedback_from_server(self):
...
+                        feedback_data = response.json()
+                        if feedback_data == {}:
+                            print("Query Client: Waiting for feedback...")
+                            continue
+                        else: 
+                            print("Query Client: Feedback received")
+                            break
+                except requests.exceptions.Timeout:
+                    print("Query Client: Timeout error occurred while waiting for feedback.")
+                    raise requests.exceptions.Timeout
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:184]: Similarly here: If you give the exception a name `except ... as exc` you can later re-raise it with `raise exc`.

File: prefq/examples/imitation_preference_comparisons.py [hunk 4]
```diff
@@ -167,16 +168,23 @@ def _get_feedback_from_server(self):
...
+                except requests.exceptions.RequestException as e:
+                    print(f"Query Client: An error occurred while waiting for feedback: {e}")
+                    raise requests.exceptions.RequestException
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:187]: Same here: Re-raise exception instead of recreating it.

File: prefq/examples/imitation_preference_comparisons.py [hunk 5]
```diff
@@ -74,16 +75,35 @@ def gather(self) -> Tuple[Sequence[TrajectoryWithRewPair], np.ndarray]:
...
+
+            # Order of queries and preferences must match, but the server response is unordered
+            # due to:
+            #   (1) The asynchronous nature feedback collection (multiple users can send feedback)
+            #   (2) The nature of the flask.jsonify() function, that orders keys alphanumerically.
+            #       This behavior can be turned off, but makes no sense due to (1)
+            query_dict[query_id] = is_left_preferred
+
+        print("\n\nOrdered Query Response")
+        for query_id, is_left_preferred in query_dict.items():
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:97]: We should not rely on the order of a dictionary. Although dictionaries technically have an order, it is not really intended to be an ordered datatype and in my opinion it is not the best style to rely on it.

I propose to just keep two dictionaries, one for queries and one for labels. Both are indexed by query IDs. In the end, we can iterate over all query IDs and create two list (`queries` and `preferences`). Something like

```python
queries = []
preferences = []
for query_id in feedback_data.keys():
    queries.append(pending_queries[query_id])
    preferences.append(feedback_data[query_id])
```

**Review by timokau — CHANGES_REQUESTED** [review]
Looks pretty good already! I left some comments for minor things to improve.


Regarding the crash/caching issue, we should determine if it is reproducible (under linux and/or WSL) and whether it is caused by this PR or already happened earlier.

File: prefq/examples/imitation_preference_comparisons.py [hunk 6]
```diff
@@ -172,52 +188,73 @@ def _wait_for_feedback_request(url):
...
-    preference_gatherer=gatherer,
-    reward_trainer=reward_trainer,
-    initial_epoch_multiplier=1,
-)
-
-pref_comparisons.train(total_timesteps=5_000, total_comparisons=200)
-
-reward, _ = evaluate_policy(agent.policy, venv, 10)
-print("Reward:", reward)
+class ContextManager():
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:191]: We should give this context manager a more descriptive name (`EnvClosingContext`).

File: prefq/examples/imitation_preference_comparisons.py [hunk 7]
```diff
@@ -172,52 +188,73 @@ def _wait_for_feedback_request(url):
...
+    def __init__(self, env):
+        self.env = env
+
+    def __enter__(self):
+        print("entering")
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:196]:
```suggestion
        pass
```

File: prefq/examples/imitation_preference_comparisons.py [hunk 8]
```diff
@@ -172,52 +188,73 @@ def _wait_for_feedback_request(url):
...
+
+    def __exit__(self, type ,value ,traceback):
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:198]:
```suggestion
    def __exit__(self, type, value, traceback):
```

File: prefq/examples/imitation_preference_comparisons.py [hunk 9]
```diff
@@ -172,52 +188,73 @@ def _wait_for_feedback_request(url):
...
+        self.env.close()
+
+with ContextManager(venv):
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:201]: Please add a comment that explains why we need this context manager, i.e., that it ensures the cached trajectories are cleaned up (which happens when the environment is closed).

File: prefq/examples/imitation_preference_comparisons.py [hunk 10]
```diff
@@ -172,52 +188,73 @@ def _wait_for_feedback_request(url):
...
+        reward_trainer=reward_trainer,
+        initial_epoch_multiplier=1,
+    )
+
+    pref_comparisons.train(total_timesteps=5_000, total_comparisons=200)
+
+    reward, _ = evaluate_policy(agent.policy, venv, 10)
+    print("Reward:", reward)
+
+    venv.close()
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:260]:
```suggestion
```

This line is redundant with the context manager.

**Review by timokau — CHANGES_REQUESTED** [review]
Looks good, just some tiny changes necessary before this is ready for merge.

**Review by timokau — APPROVED** [review]
Thank you! Great work, this marks the milestone #25 as completed :) :tada:
